### DISCUSSION TITLE
Architecture changes regarding python-eq3bt

### DISCUSSION THREAD

**Discussion by EuleMitKeule** [dbuezas/eq3btsmart#87] [General]
This Home Assistant integration currently contains an altered version of the `python-eq3bt` package. Having the goal in mind to get this integration merged into the Home Assistant core I think it will be necessary to restructure this architecture in one of the following ways:

### 1. **Merge into `rytilahti/python-eq3bt`**
* Open a PR to [https://github.com/rytilahti/python-eq3bt](https://github.com/rytilahti/python-eq3bt) and get the changes to the package merged.
* Depend on the `python-eq3bt` package.
* Only implement the Home Assistant integration in this repository.
  * **Problem:** The changed package currently depends on the `homeassistant` package; it would not make much sense to distribute this as a standalone library.
  * **Problem:** `python-eq3bt` seems to be unmaintained.

### 2. **Create and maintain a fork of `rytilahti/python-eq3bt`**
* We could create and maintain a fork or copy of the original package and distribute it as `python-eq3bt2` or similar.
* Remove the `homeassistant` dependency from that package for it to make sense as a standalone library.
  * **Problem:** Large changes would have to be made to the package and the integration to remove the `homeassistant` dependency.
  * **Problem:** Just another repository to maintain.

### 3. **Unify package and integration**
* Restructure the integration code to simply include the `BleakConnection` and `Thermostat` classes.
* This is essentially what is already done but would make it more "official."
* **Problem:** The integration might be too complicated to be merged into `core`, and Home Assistant might require us to keep the connection logic separate from the integration.

Personally I would prefer to include the connection logic within the integration as stated in the third proposal.
Since the connection logic currently depends strongly on the Home Assistant `bluetooth` component, I feel that this would be acceptable to be included in the Home Assistant Core. However we should take a look at how other integrations that are using bluetooth handle this kind of problem.

**Comment by dbuezas** [comment]
Hi again!

## 1. Merge into rytilahti/python-eq3bt

This repo diverged too much and also removed features irrelevant to home assistant, plus the owner doesn't own an eq3 anymore (I know because I bought it from him) so I doubt this would work out.

## 2. Create and maintain a fork of rytilahti/python-eq3bt

Way to go! We need to publish it to a package repository too

## 3. Unify package and integration

I believe Home Assistant has a policy against this.

**Reply by EuleMitKeule** [reply]
Would you say that the dependency of `python-eq3bt` towards the Home Assistant bluetooth component would need to be removed in order to isolate it into a separate package? I think the bluetooth component code that is used to get the device connection only works when Home Assistant is running, so I would say it wouldn't make that much sense to publish this as a standalone package in its current state. The code could be refactored though in such a way that the `eq3btsmart` integration passes the required methods or objects to `python-eq3bt` and `python-eq3bt` could implement a default way to connect to the thermostats without having the need of a running Home Assistant installation. What are your thoughts?
Also do you want to create the new repository containing `python-eq3bt2`? I will have some time to work on both projects in the next few weeks, so if you want, I could also "takeover" the `python-eq3bt` part to my account.

Oh, and should the new repository be a fork of `python-eq3bt` or be its own thing? I would vouch for the latter, while of course still including the original license for copyright reasons.

Another idea could be to keep both the package and the integration in a single repository. I used this kind of setup in my repo https://github.com/eulemitkeule/estimenergy.
Once the integration part would be merged into `core` the custom component directory could then be removed from this repo.

**Reply by dbuezas** [reply]
I think they would all work. I slightly prefer keeping everything together for as long as possible so things stay simple.
Regarding repositories and ownership, I made you a contributor of this repo so you can also merge and whatnot.

If you want to fork you of course can (that's what I did myself when the original stopped working), but my preference is to work on a branch of this repo.

One consideration is that I'm not sure if the HA team will like the "hacky" parts like selecting the specific Bluetooth dongle for each eq3. I use that myself to overcome the maximum connected devices per dongle limitation. Since the valves require pairing, leaving this to HA doesn't work.

**Reply by EuleMitKeule** [reply]
Thank you for giving me push access! I already added some of the tools I recommended in #88 and began restructuring the repo and fixing some code style and typing related issues. For now I will only merge these changes into the `develop` branch until everything is thoroughly tested by us and approved by you. 

I have also added a license file (MIT license) with both our names and also kept the license of the original `python-eq3bt` package.

For testing purposes I already built and published the [`eq3btsmart` package](https://pypi.org/project/eq3btsmart/) to pyPI. If you tell me your username on pyPI I can add you as a co-owner of the project.

**Reply by EuleMitKeule** [reply]
I would say the next steps would be the following:

* Remove the dependency on `homeassistant` in `eq3btsmart`
* Refactor the integration to work with the changed library
* Automate the building and publishing process

Furthermore I'd like to propose renaming the integration part from `dbuezas_eq3btsmart` to either `eq3btsmart_hass` or `eq3btsmart_integration` or something similar. I think this would make the code a little bit clearer to understand and easier to distinguish between the library and integration. What are your thoughts on that?

Oh and if you'd like me to change anything about the license or something, please let me know!

**Reply by dbuezas** [reply]
Sounds good, now that the old core integration was removed, I think we can call it eq3btsmart as the original.
Re license, anything that maintains both the open source spirit, and that of the original work by @rytilahti works for me.

**Reply by dbuezas** [reply]
Calling it eq3btsmart would make the brands logo work

**Reply by EuleMitKeule** [reply]
Actually I don't think the original integration has been removed yet, it is still present as the `maxcube` integration and can be added when searching for "EQ-3" under the name "EQ-3 MAX!". Also we don't need to choose a specific name for the logo to be associated with the integration, we just have to add it to [this file](https://github.com/home-assistant/core/blob/dev/homeassistant/brands/eq3.json) when merging into `core` if I am not mistaken.
Either way naming it `eq3btsmart` is a good idea I would say. The problem is that we definetely need to choose a different name for the integration than the package or else imports will not work and dependencies would also probably be messed up. That's why I proposed naming the integration something like `eq3btsmart_hass` or `eq3btsmart_integration` because it's somewhat more specialized than the library itself.

Edit: I think I was wrong about the same name part. Since we need to use relative imports for the integration having the same name for both the integration and the library should probably fine. I also found some core integrations that depend on packages with the same name as well.

**Reply by rytilahti** [reply]
AFAIK, the linked file only handles searching inside the "add integration" dialog, but I may be wrong on that, so if it works that's great. Otherwise, you may want to look into reverting (or just re-adding) this: https://github.com/home-assistant/brands/pull/4935

**Reply by EuleMitKeule** [reply]
Oh, good to know, thanks! 
So I guess there actually was another integration present in Home Assistant that was actually named `eq3btsmart` in addition to the `maxcube` / "eQ-3 MAX!" integration with the latter still being present in `core`? All this is pretty confusing to me.

**Reply by rytilahti** [reply]
Yes, the old python-eq3bt-based integration was there since homeassistant 0.18 ([it seems :-D](https://github.com/home-assistant/home-assistant.io/blob/a1b53b172dffdf3b32104762bb94c083f5d6de24/source/_integrations/eq3btsmart.markdown)) but it was very, very outdated prior @dbuezas taking over and creating this custom integration.

Note that when you submit the integration for inclusion, do not forget to remove the redirection added in  https://github.com/home-assistant/home-assistant.io/pull/27848/files .

**Comment by rytilahti** [comment]
Glad to see that you are working to get this integrated into the core! :100: As @dbuezas mentioned, I don't have these thermostats anymore, so the python-eq3bt maintenance will be kept from my side at bare minimums (i.e., just fixing issues/merging PRs/making releases if something comes up). 

I personally also think that it doesn't make sense to try to merge back into the python-eq3bt, assuming the end goal is to mainly support the homeassistant integration. If you wish to do so though, I can move the ownership of that repository to you somehow. However, having a clean fork is probably better and less hassle for everyone.

**Reply by EuleMitKeule** [reply]
Thank you for the support and for the offer to transfer ownership! 
I agree with you that having two separate projects is probably better in this case, since the `eq3btsmart` library was altered so much by @dbuezas to depend on the Home Assistant bluetooth component and will now be changed again to remove this dependency but still work with Home Assistant. And I think having a minimal library that only contains basic connection logic and structs to communicate with the devices and not being intended as a standalone utility will probably increase the chances getting this merged into `core`. 
Since this project builds so much upon your work we of course kept your license and link to the original repo. If you want anything about that handled differently please let us know!
